### PR TITLE
[Scala] Stateful class declaration and some meta scopes

### PR DESCRIPTION
**Pull request by djspiewak**
Scala has a very rich class declaration syntax with some constraints (such as you cannot use `with` before `extends`, and cannot use multiple `extends`).  For this reason, we have long punted on this and just popped out as soon as we saw those sorts of contexts, which led to certain things being highlighted sub-optimally.  Additionally, it outright prevented ever accurately assigning the `meta.class.body` scope.

This is now fixed, and the meta scope is applied.  Additionally, support has been added for flagging certain completely-invalid syntaxes as `invalid`.  I additionally fixed some low-hanging meta scoping in other areas.  Lots more work to do here.

### CODE REVIEW

File: Scala/Scala.sublime-syntax [hunk 1]
```diff
@@ -463,14 +473,131 @@ contexts:
...
+  class-inheritance-extends-newline:
+    - include: decl-newline-double-check
+    - match: '(?=\S)'
+      set: class-inheritance
+
+  class-inheritance-extends-token:
+    - match: '{{id}}'
+      scope: entity.other.inherited-class.scala
+      set: class-inheritance-extends-token-after
+    - match: \(
```
Comment (wbond) [inline Scala/Scala.sublime-syntax]: These should have a `punctuation` scope

File: Scala/Scala.sublime-syntax [hunk 2]
```diff
@@ -463,14 +473,131 @@ contexts:
...
+    - match: '\n'
+      set: class-inheritance-extends-token-newline
+
+  class-inheritance-extends-token-newline:
+    - include: decl-newline-double-check
+    - match: '(?=\S)'
+      set: class-inheritance-extends-token
+
+  class-inheritance-extends-token-after:
+    - match: \(
```
Comment (wbond) [inline Scala/Scala.sublime-syntax]: Also `punctuation`

File: Scala/Scala.sublime-syntax [hunk 3]
```diff
@@ -463,14 +473,131 @@ contexts:
...
+      push:
+        - match: \)
+          pop: true
+        - include: main
+    - match: \[
```
Comment (wbond) [inline Scala/Scala.sublime-syntax]: Also `punctuation`

File: Scala/Scala.sublime-syntax [hunk 4]
```diff
@@ -463,14 +473,131 @@ contexts:
...
+    - match: '\n'
+      set: class-inheritance-extends-token-after-newline
+
+  class-inheritance-extends-token-after-newline:
+    - include: decl-newline-double-check
+    - match: '(?=\S)'
+      set: class-inheritance-extends-token-after
+
+  class-inheritance-early-initializer:
+    - match: \{
```
Comment (wbond) [inline Scala/Scala.sublime-syntax]: `punctuation` scope

File: Scala/Scala.sublime-syntax [hunk 5]
```diff
@@ -463,14 +473,131 @@ contexts:
...
+    - match: '(?=\{)'
+      set: class-pre-inheritance-early-initializer
 
   class-parameter-list-newline:
     - include: decl-newline-double-check
     - match: '(?=\S)'
       set: class-parameter-list
 
+  class-pre-inheritance-early-initializer:
+    - match: \{
```
Comment (wbond) [inline Scala/Scala.sublime-syntax]: `punctuation` scope

**Comment by djspiewak** [comment]
@wbond I believe I have resolved all of the appropriate changes.

**Pull request merged**